### ISSUE TITLE
GC sweep design: client session expiry to get rid of in-memory unreferenced content

### ISSUE THREAD

**Issue by agarwal-navin** [microsoft/FluidFramework#7491]
When unreferenced content is deleted by GC, it needs to ensure that the content cannot be revived. For this, we want to reload client sessions that may have in-memory referenced to content that is about to be deleted. For example, clients may have stored these in their undo-redo or cut-copy-paste stack.
Basically, if a client had access to the content before it was unreferenced, it could have held an in-memory reference. If we reload the session or force to be read-only, the in-memory content will be flushed and it would be safe to delete the content.
There are 2 different approaches to this with their pros and cons:

Approach 1
------------
In this approach, every client session is expired after a specific time (maxSessionActiveTime). maxSessionActiveTime is calculated from the timestamp of the summary (summaryTimestamp) that the client loads from.
1. When a container loads, it uses the timestamp in the IConnected message (connectedTimestamp) and the summaryTimestamp to start the session expiry timer.
  `sessionExpiryTime = maxSessionActiveTime - (connectedTimestamp - summaryTimestamp)`
2. If the summary that a client loads from is older than maxSessionActiveTime, expire the session immediately. Else, start a timer for sessionExpiryTime and then expire the session.
3. Add 2 days to account for snapshot caching. We need to enforce this in the driver such that a cached snapshot cannot be used after 2 days.
4. Delete content physically only after it was unreferenced for at least maxSessionActiveTime + 2 days.

One problem with this approach is how to identify the scenario when a client loads from a summary that is > maxSessionActiveTime old and that is only snapshot available. Or it's a cached snapshot and there are newer snapshots available but the client doesn't know about it yet. We cannot expire the session in this case.
This problem can be solved by the # 3 above. The core issue that we are trying to solve is that the client can hold in-memory copy of objects that are referenced when it loads and the object becomes unreferenced later in its lifetime. If we limit cached snapshot usage to 2 days, then all objects will fall into one of these categories:
- They are unreferenced in the base snapshot that the client loads from.
- They are referenced in the base snapshot and becomes unreferenced after the client loads.
- They are referenced in the base snapshot and became unreferenced in a snapshot in the past 2 days (since a newer snapshot could only have been posted in that time due to # 3 above).
By taking this into account, if we expire this client after maxSessionActiveTime from when it loads and delete content only after > maxSessionActiveTime + 2 days have passed, we can be sure that in-memory objects are flushed. We also don't really need # 1 above, we can simply always start a setTimeout for maxSessionActiveTime when a client loads.

The other cons of this approach is that every client is responsible for maintaining its own timer, making this logic distribued in all clients. Every client will be forced to expire and reload irrespective of whether it could have in-memory content or not. Lastly, because of # 3 above, the policy is split across layers which is not very ideal.

Approach 2
----------
This approach introduces a "delete" op that will be sent out when an object is identified as deletable. It solves the cons of Approach 1 but it is a little more complex and requires additional functionality.
1. When GC runs, it identifies objects that can be deleted. Basically, objects that have been deleted for more than a specific amount of time (maxUnreferencedDuration).
2. It stores the ids of these objects in memory, uploads a summary and waits for it to be ack'd.
3. When the summary is ack'd, it generates a delete op with the ids of the objects in # 2 above. Waiting for the summary ack before sending the delete op ensures that there is always a summary before the delete op is sent. This will become relevant later.
4. When the summarizer receives this delete op, it updates the deleted objects with the seq# of the delete op.
5. When other clients get the delete op, they check whether any of the deleted data stores is referenced. If so, this client's session is expired as it may have stored in-memory copy to these objects. If all the objects are unreferenced, it doesn't need to be expired.
6. When the summarizer client sees that the MSN has moved past the seq# in # 4 above, it deletes these objects as the clients that can have in-memory copies would have been expired.

This approach ensures that clients only reload if they could have had in-memory copies of objects being deleted. It also handles the problem of clients loading from snapshot >maxSessionActiveTime because they won't have to expire unless they see the delete op which should happen after at least maxUnreferencedDuration time.

One problem with this approach is that when a client expires and has to reload when it sees the delete op, how do we ensure it loads from a newer summary (because it may have cached snapshot). The solution can be to invalidate the cache when we expire the client so that it has to fetch a new snapshot. From # 3 above, we make sure that there is a newer snapshot available when the delete op is generated. We already have the logic of invalidating the cache when epoch failures occur; we can incorporate this case in that logic.

I like approach 2 as it is uses ops which make it more reliable and it has the core logic contained in the summarizer client. However, approach 1 is simpler and should work well for now. So, I propose that we go ahead with approach 1 and monitor if it is good enough. If not, we can fallback and implement approach 2 later.

**Comment by agarwal-navin** [comment]
@anthony-murphy @vladsud @curtisman @andre4i @pleath FYI

**Comment by agarwal-navin** [comment]
There is another issue which exists irrespective of which approach we choose as described here - https://github.com/microsoft/FluidFramework/issues/7493

**Comment by agarwal-navin** [comment]
@vladsud @anthony-murphy Ping
